cf_xarray/geometry.py

```diff
@@ -0,0 +1,267 @@
+
+    This function makes sense only if the dimension of the geometry variable has no coordinate,
+    or if that coordinate has repeated values for each geometry.
+
+    Parameters
+    ----------
+    ds : xr.Dataset
+      A Dataset.
+    geom_var : string
+      Name of the variable in `ds` that contains the geometry objects (shapely.geometry). The variable must be 1D.
+    new_dim : string
+      Name of the new dimension.
+
+    Returns
+    -------
+    Dataset
+      All variables sharing the dimension of `ds[geom_var]` are reshaped so that `new_dim` as a length equal to the
+      number of unique geometries.
+    """
+    if ds[geom_var].ndim > 1:
+        raise ValueError(
+            f"The geometry variable must be 1D. Got ds[{geom_var}] with dims {ds[geom_var].dims}."
+        )
+
+    # Shapely objects are not hashable, thus np.unique cannot be used directly.
+    # This trick is stolen from geopandas.
+    _, unique_indexes, inv_indexes = np.unique(
+        [g.wkb for g in ds[geom_var].values], return_index=True, return_inverse=True
+    )
+    old_name = ds[geom_var].dims[0]
+
+    if old_name in ds.coords:
+        old_values = ds[old_name].values
+    else:
+        # A dummy coord, a kind of counter, independent for each unique geometries
+        old_values = np.array(
+            [(inv_indexes[:i] == ind).sum() for i, ind in enumerate(inv_indexes)]
+        )
+
+    multi_index = pd.MultiIndex.from_arrays(
+        (inv_indexes, old_values), names=(new_dim, old_name)
+    )
+    temp_name = xr.core.utils.get_temp_dimname(ds, "multi_index")
+    out = ds.rename({old_name: temp_name})
+    out[temp_name] = multi_index
+    out = out.unstack(temp_name)
+
+    # geom_var was reshaped also, reconstruct it from the unique values.
+    unique_indexes = xr.DataArray(unique_indexes, dims=(new_dim,))
+    out[geom_var] = ds[geom_var].isel({old_name: unique_indexes})
+    if old_name not in ds.coords:
+        # If there was no coord before, drop the dummy one we made.
+        out = out.drop_vars(old_name)
+    return out
+
+
+def geometry_to_cf(geometries: Union[xr.DataArray, Sequence], grid_mapping: str = None):
+    """Convert a DataArray with shapely geometry objects into a CF-compliant dataset.
+
+    WARNING: Only point geometries are currently implemented.
+
+    Parameters
+    ----------
+    geometries : sequence of shapely geometries or xarray.Dataset
+      A sequence of geometry objects or a Dataset with a "geometry" variable storing such geometries.
+      All geometries must be of the same base type : Point, Line or Polygon, but multipart geometries are accepted.
+    grid_mapping : str, optional
+      A CF grid mapping name. When given, coordinates and attributes are named and set accordingly.
+      Defaults to None, in which case the coordinates are simply names "crd_x" and "crd_y".
+      WARNING: only the `longitude_latitude` grid mapping is currently implemented.
```

Comment:
```suggestion
      WARNING: only the `longitude_latitude` grid mapping is currently implemented.
```

same for this warning